DRA scheduler: flaky TestPlugin/extended-resource-name-no-resource

### Which jobs are flaking?

https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-unit

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/137108/pull-kubernetes-unit/2024138310519099392


### Which tests are flaking?

k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources, TestPlugin/extended-resource-name-no-resource

### Since when has it been flaking?

Unknown, only seen once in a presubmit.

### Testgrid link

_No response_

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

Can be reproduced locally:

```
$ go test -c ./pkg/scheduler/framework/plugins/dynamicresources/
$ stress -p 512 ./dynamicresources.test -test.run=TestPlugin/extended-resource-name-no-resource
5s: 2048 runs so far, 0 failures

/tmp/go-stress-20260218T163653-3909757302
E0218 16:37:01.852829  224548 reflector.go:434] "Reflector stopped with error" err="context canceled" type="*v1.ResourceClaim" reflector="k8s.io/client-go/informers/factory.go:177"
E0218 16:37:01.852844  224548 reflector.go:434] "Reflector stopped with error" err="context canceled" type="*v1.DeviceClass" reflector="k8s.io/client-go/informers/factory.go:177"
E0218 16:37:01.852853  224548 reflector.go:434] "Reflector stopped with error" err="context canceled" type="*v1alpha3.DeviceTaintRule" reflector="k8s.io/client-go/informers/factory.go:177"
E0218 16:37:01.852860  224548 reflector.go:434] "Reflector stopped with error" err="context canceled" type="*v1.ResourceSlice" reflector="k8s.io/client-go/informers/factory.go:177"
--- FAIL: TestPlugin (1.00s)
    --- FAIL: TestPlugin/extended-resource-name-no-resource (1.00s)
        feature_gate.go:170: I0218 16:37:01.850093] I0218 16:37:01.850067] Updated featureGates={"DRAExtendedResource":true}
        shared_informer.go:401: I0218 16:37:01.851141] Waiting for caches to sync
        shared_informer.go:408: I0218 16:37:01.852368] Caches are synced
        --- FAIL: TestPlugin/extended-resource-name-no-resource/prefilter (0.00s)
            dynamicresources_test.go:2568: ERROR: I0218 16:37:01.852528]
                	Error Trace:	/nvme/gopath/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:2734
                		            				/nvme/gopath/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:2568
                		            				/nvme/gopath/src/k8s.io/kubernetes/test/utils/ktesting/tcontext.go:280
                		Error:      	Expected nil, but got: &framework.Status{code:5, reasons:[]string(nil), err:error(nil), plugin:""}
                		Test:       	TestPlugin/extended-resource-name-no-resource/prefilter
        --- FAIL: TestPlugin/extended-resource-name-no-resource/filter/worker (0.00s)
```


### Relevant SIG(s)

/sig scheduling
/wg device-management
/assign @yliaog 
/cc @bart0sh

--- a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
+++ b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
@@ -2997,6 +2997,7 @@ func setup(tCtx ktesting.TContext, args *config.DynamicResourcesArgs, nodes []*v
 	tc.client.ReactionChain = append(apiReactors, tc.client.ReactionChain...)
 
 	tc.informerFactory = informers.NewSharedInformerFactory(tc.client, 0)
+	var doneCheckers []cache.DoneChecker
 	resourceSliceTrackerOpts := resourceslicetracker.Options{
 		EnableDeviceTaintRules: true,
 		SliceInformer:          tc.informerFactory.Resource().V1().ResourceSlices(),
@@ -3006,6 +3007,7 @@ func setup(tCtx ktesting.TContext, args *config.DynamicResourcesArgs, nodes []*v
 	}
 	resourceSliceTracker, err := resourceslicetracker.StartTracker(tCtx, resourceSliceTrackerOpts)
 	require.NoError(tCtx, err, "couldn't start resource slice tracker")
+	doneCheckers = append(doneCheckers, resourceSliceTracker.HasSyncedChecker())
 
 	claimsCache := assumecache.NewAssumeCache(tCtx.Logger(), tc.informerFactory.Resource().V1().ResourceClaims().Informer(), "resource claim", "", nil)
 	// NewAssumeCache calls the informer's AddEventHandler method to register
@@ -3018,14 +3020,14 @@ func setup(tCtx ktesting.TContext, args *config.DynamicResourcesArgs, nodes []*v
 	// This is not the registered handler that is used by the DRA
 	// manager, but it is close enough because the assume cache
 	// uses a single boolean for "is synced" for all handlers.
-	registeredHandler := claimsCache.AddEventHandler(cache.ResourceEventHandlerFuncs{})
+	doneCheckers = append(doneCheckers, claimsCache.AddEventHandler(cache.ResourceEventHandlerFuncs{}).HasSyncedChecker())
 
 	tc.draManager = NewDRAManager(tCtx, claimsCache, resourceSliceTracker, tc.informerFactory)
 	if features.EnableDRAExtendedResource {
 		cache := tc.draManager.DeviceClassResolver().(*extendedresourcecache.ExtendedResourceCache)
-		if _, err := tc.informerFactory.Resource().V1().DeviceClasses().Informer().AddEventHandler(cache); err != nil {
-			tCtx.Logger().Error(err, "failed to add device class informer event handler")
-		}
+		deviceClassHandlerRegistration, err := tc.informerFactory.Resource().V1().DeviceClasses().Informer().AddEventHandler(cache)
+		require.NoError(tCtx, err, "failed to add device class informer event handler")
+		doneCheckers = append(doneCheckers, deviceClassHandlerRegistration.HasSyncedChecker())
 	}
 
 	opts := []runtime.Option{
@@ -3068,11 +3070,11 @@ func setup(tCtx ktesting.TContext, args *config.DynamicResourcesArgs, nodes []*v
 	})
 
 	tc.informerFactory.WaitForCacheSync(tCtx.Done())
-	// The above does not tell us if the registered handler (from NewAssumeCache)
-	// is synced, we need to wait until HasSynced of the handler returns
-	// true, this ensures that the assume cache is in sync with the informer's
+	// The above does not tell us if the registered handlers (e.g. from NewAssumeCache)
+	// are synced, we need to wait until the event handlers confirm that they are synced.
+	// This ensures that the assume cache is in sync with the informer's
 	// store which has been informed by at least one full LIST of the underlying storage.
-	cache.WaitForNamedCacheSyncWithContext(tCtx, registeredHandler.HasSynced, resourceSliceTracker.HasSynced)
+	cache.WaitFor(tCtx, "event handlers", doneCheckers...)
 
 	for _, node := range nodes {
 		nodeInfo := framework.NewNodeInfo()

